src/stream/src/executor/backfill/cdc/upstream_table/snapshot.rs

```diff
@@ -128,40 +113,147 @@ impl UpstreamTableRead for UpstreamTableReader<ExternalStorageTable> {
             )
         });
 
-        #[for_await]
-        for chunk in chunk_stream {
-            let chunk = chunk?;
-            let chunk_size = chunk.capacity();
+        let mut read_args = args;
+        // loop to read all data from the table
+        loop {
+            tracing::debug!(
+                "snapshot_read primary keys: {:?}, current_pos: {:?}",
+                primary_keys,
+                read_args.current_pos
+            );
+
+            let mut read_count: usize = 0;
+            let row_stream = self.inner.table_reader().snapshot_read(
+                self.inner.schema_table_name(),
+                read_args.current_pos.clone(),
+                primary_keys.clone(),
+                batch_size,
+            );
 
-            if args.rate_limit_rps.is_none() || chunk_size == 0 {
-                // no limit, or empty chunk
-                yield Some(chunk);
-                continue;
+            pin_mut!(row_stream);
+            let mut builder = DataChunkBuilder::new(
+                self.inner.schema().data_types(),
+                limited_chunk_size(read_args.rate_limit_rps),
+            );
+            let chunk_stream = iter_chunks(row_stream, &mut builder);
+            let mut current_pk_pos = read_args.current_pos.clone().unwrap_or_default();
+
+            #[for_await]
+            for chunk in chunk_stream {
+                let chunk = chunk?;
+                let chunk_size = chunk.capacity();
+                read_count += chunk.cardinality();
+                current_pk_pos = get_new_pos(&chunk, &read_args.pk_in_output_indices);
+
+                if read_args.rate_limit_rps.is_none() || chunk_size == 0 {
+                    // no limit, or empty chunk
+                    yield Some(chunk);
+                    continue;
+                } else {
+                    // Apply rate limit, see `risingwave_stream::executor::source::apply_rate_limit` for more.
+                    // May be should be refactored to a common function later.
+                    let limiter = limiter.as_ref().unwrap();
+                    let limit = read_args.rate_limit_rps.unwrap() as usize;
+
+                    // Because we produce chunks with limited-sized data chunk builder and all rows
+                    // are `Insert`s, the chunk size should never exceed the limit.
+                    assert!(chunk_size <= limit);
```

Comment:
If the last event is `Update` i.e. (an `U-` row followed by an `U+` row), the chunk_size could be `limit + 1`

https://github.com/risingwavelabs/risingwave/blob/75ebd1ea43b211b16321b444b6d47c0ed039a539/src/common/src/array/stream_chunk.rs#L181-L182

Reply:
This part is a merge with main. And for the snapshot read in cdc backfill, stream chunk only contain `Insert` event, there is no `Update`.